src/libos/src/signal/do_sigreturn.rs

```diff
@@ -278,29 +289,87 @@ fn handle_signals_by_user(
-            let fpregs = FpRegs::save();
-            ucontext.fpregs.copy_from_slice(fpregs.as_slice());
+            // curr_user_ctxt.extra_context_ptr != ptr::null_mut()
+            match curr_user_ctxt.extra_context {
+                ExtraContext::Fpregs => {
+                    // When sysret, use fpregs context from the latest signal handler. Thus free old fpregs to
+                    // prevent memory leakage.
+                    drop(unsafe { Box::from_raw(curr_user_ctxt.extra_context_ptr as *mut FpRegs) });
+                    let fpregs = Box::new(FpRegs::save());
+                    ucontext.fpregs.copy_from_slice(fpregs.as_slice());
+                    ucontext.uc_mcontext.inner.extra_context_ptr = Box::into_raw(fpregs) as *mut u8;
+                    ucontext.uc_mcontext.inner.extra_context = ExtraContext::Fpregs;
+                    ucontext.uc_mcontext.inner.extra_context_size =
+                        std::mem::size_of::<FpRegs>() as u64;
+                }
+                // force_signal from exception or deliver_signal when interrupt returns
+                ExtraContext::XsaveOnStack => {
+                    // Save Xsave area at the heap and store the pointer to Occlum's defined CpuContext.
+                    // So that recursive exception will not overwrite the xsave area.
+                    let xsave_size = curr_user_ctxt.extra_context_size;
+                    debug_assert!(
+                        xsave_size != 0 && curr_user_ctxt.extra_context_ptr != ptr::null_mut()
+                    );
+                    let mut xsave_area =
+                        ManuallyDrop::new(BoxXsaveArea::new_uninit(xsave_size as usize));
+                    xsave_area.as_mut_slice().copy_from_slice(unsafe {
+                        std::slice::from_raw_parts(
+                            curr_user_ctxt.extra_context_ptr,
+                            xsave_size as usize,
+                        )
+                    });
+                    ucontext.uc_mcontext.inner.extra_context_ptr = xsave_area.ptr();
+                    ucontext.uc_mcontext.inner.extra_context = ExtraContext::XsaveOnHeap;
+                    // extra_context_size is the same
+
+                    // Copy fpregs to ucontext for user's handler
+                    let fpregs_area = xsave_area.get_fpregs();
+                    ucontext.fpregs.copy_from_slice(fpregs_area.as_slice());
+                }
+                // Handle another signal before sysret during exception/interrupt
+                ExtraContext::XsaveOnHeap => {
+                    debug_assert!(
+                        curr_user_ctxt.extra_context_size != 0
+                            && curr_user_ctxt.extra_context_ptr != ptr::null_mut()
+                    );
+                    // xsave area is already saved on the heap. Just reuse it.
+                    ucontext.uc_mcontext.inner.extra_context_ptr = curr_user_ctxt.extra_context_ptr;
```

Comment:
I think resetting `ucontext.uc_mcontext.inner.extra_context_ptr` is redundant.

Reply:
Removed.